// @antv/l7-map/lib/handler/click_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
// @ts-ignore

class ClickZoomHandler {
  constructor() {
    (0, _defineProperty2.default)(this, "enabled", void 0);
    (0, _defineProperty2.default)(this, "active", void 0);
    this.reset();
  }
  reset() {
    this.active = false;
  }
  dblclick(e, point) {
    e.preventDefault();
    return {
      cameraAnimation: map => {
        map.easeTo({
          duration: 300,
          zoom: map.getZoom() + (e.shiftKey ? -1 : 1),
          around: map.unproject(point)
        }, {
          originalEvent: e
        });
      }
    };
  }
  enable() {
    this.enabled = true;
  }
  disable() {
    this.enabled = false;
    this.reset();
  }
  isEnabled() {
    return this.enabled;
  }
  isActive() {
    return this.active;
  }
}
exports.default = ClickZoomHandler;
