// @antv/l7-maps/es/mapbox/Viewport.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import WebMercatorViewport from 'viewport-mercator-project';
export default class Viewport {
  constructor() {
    _defineProperty(this, "viewport", void 0);
  }
  syncWithMapCamera(mapCamera) {
    const {
      center,
      zoom,
      pitch,
      bearing,
      viewportHeight,
      viewportWidth
    } = mapCamera;

    /**
     * Deck.gl 使用的也是 Mapbox 同步相机，相机参数保持一致
     * 例如相机高度固定为 height * 1.5，因此不需要传
     */
    this.viewport = new WebMercatorViewport({
      width: viewportWidth,
      height: viewportHeight,
      longitude: center && center[0],
      latitude: center && center[1],
      zoom,
      pitch,
      bearing
    });
  }
  getZoom() {
    return this.viewport.zoom;
  }
  getZoomScale() {
    return Math.pow(2, this.getZoom());
  }
  getCenter() {
    return [this.viewport.longitude, this.viewport.latitude];
  }
  getProjectionMatrix() {
    return this.viewport.projectionMatrix;
  }
  getModelMatrix() {
    return [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1];
  }
  getViewMatrix() {
    return this.viewport.viewMatrix;
  }
  getViewMatrixUncentered() {
    // @ts-ignore
    return this.viewport.viewMatrixUncentered;
  }
  getViewProjectionMatrix() {
    // @ts-ignore
    return this.viewport.viewProjectionMatrix;
  }
  getViewProjectionMatrixUncentered() {
    // @ts-ignore
    return this.viewport.viewProjectionMatrix;
  }
  getFocalDistance() {
    return 1;
  }
  projectFlat(lngLat, scale) {
    return this.viewport.projectFlat(lngLat, scale);
  }
}
