// @antv/l7-map/lib/geo/lng_lat.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.earthRadius = exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _util = require("../util");
const earthRadius = exports.earthRadius = 6371008.8;
class LngLat {
  static convert(input) {
    if (input instanceof LngLat) {
      return input;
    }
    if (Array.isArray(input) && (input.length === 2 || input.length === 3)) {
      return new LngLat(Number(input[0]), Number(input[1]));
    }
    if (!Array.isArray(input) && typeof input === 'object' && input !== null) {
      const lng = 'lng' in input ? input.lng : input.lon;
      return new LngLat(
      // flow can't refine this to have one of lng or lat, so we have to cast to any
      Number(lng), Number(input.lat));
    }
    throw new Error('`LngLatLike` argument must be specified as a LngLat instance, an object {lng: <lng>, lat: <lat>}, an object {lon: <lng>, lat: <lat>}, or an array of [<lng>, <lat>]');
  }
  constructor(lng, lat) {
    (0, _defineProperty2.default)(this, "lng", void 0);
    (0, _defineProperty2.default)(this, "lat", void 0);
    if (isNaN(lng) || isNaN(lat)) {
      throw new Error(`Invalid LngLat object: (${lng}, ${lat})`);
    }
    this.lng = +lng;
    this.lat = +lat;
    if (this.lat > 90 || this.lat < -90) {
      throw new Error('Invalid LngLat latitude value: must be between -90 and 90');
    }
  }
  wrap() {
    return new LngLat((0, _util.wrap)(this.lng, -180, 180), this.lat);
  }
  toArray() {
    return [this.lng, this.lat];
  }
  // public toBounds(radius: number = 0) {
  //   const earthCircumferenceInMetersAtEquator = 40075017;
  //   const latAccuracy = (360 * radius) / earthCircumferenceInMetersAtEquator;
  //   const lngAccuracy = latAccuracy / Math.cos((Math.PI / 180) * this.lat);

  //   return new LngLatBounds(
  //     new LngLat(this.lng - lngAccuracy, this.lat - latAccuracy),
  //     new LngLat(this.lng + lngAccuracy, this.lat + latAccuracy),
  //   );
  // }
  toString() {
    return `LngLat(${this.lng}, ${this.lat})`;
  }
  distanceTo(lngLat) {
    const rad = Math.PI / 180;
    const lat1 = this.lat * rad;
    const lat2 = lngLat.lat * rad;
    const a = Math.sin(lat1) * Math.sin(lat2) + Math.cos(lat1) * Math.cos(lat2) * Math.cos((lngLat.lng - this.lng) * rad);
    const maxMeters = earthRadius * Math.acos(Math.min(a, 1));
    return maxMeters;
  }
}
exports.default = LngLat;
