// @antv/l7-map/es/handler/tap/single_tap_recognizer.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
// @ts-ignore
import Point from "../../geo/point";
import { indexTouches } from "../handler_util";
function getCentroid(points) {
  const sum = new Point(0, 0);
  for (const point of points) {
    sum._add(point);
  }
  // @ts-ignore
  return sum.div(points.length);
}
export const MAX_TAP_INTERVAL = 500;
export const MAX_TOUCH_TIME = 500;
export const MAX_DIST = 30;
export default class SingleTapRecognizer {
  constructor(options) {
    _defineProperty(this, "numTouches", void 0);
    _defineProperty(this, "centroid", void 0);
    _defineProperty(this, "startTime", void 0);
    _defineProperty(this, "aborted", void 0);
    _defineProperty(this, "touches", void 0);
    this.reset();
    this.numTouches = options.numTouches;
  }
  reset() {
    // @ts-ignore
    delete this.centroid;
    // @ts-ignore
    delete this.startTime;
    // @ts-ignore
    delete this.touches;
    this.aborted = false;
  }
  touchstart(e, points, mapTouches) {
    if (this.centroid || mapTouches.length > this.numTouches) {
      this.aborted = true;
    }
    if (this.aborted) {
      return;
    }
    if (this.startTime === undefined) {
      this.startTime = e.timeStamp;
    }
    if (mapTouches.length === this.numTouches) {
      this.centroid = getCentroid(points);
      this.touches = indexTouches(mapTouches, points);
    }
  }
  touchmove(e, points, mapTouches) {
    if (this.aborted || !this.centroid) {
      return;
    }
    const newTouches = indexTouches(mapTouches, points);
    for (const id in this.touches) {
      if (this.touches[id]) {
        const prevPos = this.touches[id];
        const pos = newTouches[id];
        if (!pos || pos.dist(prevPos) > MAX_DIST) {
          this.aborted = true;
        }
      }
    }
  }
  touchend(e, points, mapTouches) {
    if (!this.centroid || e.timeStamp - this.startTime > MAX_TOUCH_TIME) {
      this.aborted = true;
    }
    if (mapTouches.length === 0) {
      const centroid = !this.aborted && this.centroid;
      this.reset();
      if (centroid) {
        return centroid;
      }
    }
  }
}
