// @antv/l7-maps/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  MapType: true,
  GaodeMapV1: true,
  GaodeMap: true,
  GaodeMapV2: true,
  BaiduMap: true,
  Earth: true,
  GoogleMap: true,
  Map: true,
  Mapbox: true,
  MapLibre: true,
  TMap: true,
  TencentMap: true
};
Object.defineProperty(exports, "BaiduMap", {
  enumerable: true,
  get: function () {
    return _bmap.default;
  }
});
Object.defineProperty(exports, "Earth", {
  enumerable: true,
  get: function () {
    return _earth.default;
  }
});
Object.defineProperty(exports, "GaodeMap", {
  enumerable: true,
  get: function () {
    return _amap2.default;
  }
});
Object.defineProperty(exports, "GaodeMapV1", {
  enumerable: true,
  get: function () {
    return _amap.default;
  }
});
Object.defineProperty(exports, "GaodeMapV2", {
  enumerable: true,
  get: function () {
    return _amap2.default;
  }
});
Object.defineProperty(exports, "GoogleMap", {
  enumerable: true,
  get: function () {
    return _gmap.default;
  }
});
Object.defineProperty(exports, "Map", {
  enumerable: true,
  get: function () {
    return _map.default;
  }
});
Object.defineProperty(exports, "MapLibre", {
  enumerable: true,
  get: function () {
    return _maplibre.default;
  }
});
Object.defineProperty(exports, "MapType", {
  enumerable: true,
  get: function () {
    return _l7Utils.MapType;
  }
});
Object.defineProperty(exports, "Mapbox", {
  enumerable: true,
  get: function () {
    return _mapbox.default;
  }
});
Object.defineProperty(exports, "TMap", {
  enumerable: true,
  get: function () {
    return _tdtmap.default;
  }
});
Object.defineProperty(exports, "TencentMap", {
  enumerable: true,
  get: function () {
    return _tmap.default;
  }
});
var _l7Utils = require("@antv/l7-utils");
var _amap = _interopRequireDefault(require("./amap/"));
var _amap2 = _interopRequireDefault(require("./amap2/"));
var _bmap = _interopRequireDefault(require("./bmap/"));
var _earth = _interopRequireDefault(require("./earth/"));
var _gmap = _interopRequireDefault(require("./gmap"));
var _map = _interopRequireDefault(require("./map/"));
var _mapbox = _interopRequireDefault(require("./mapbox/"));
var _maplibre = _interopRequireDefault(require("./maplibre"));
var _tdtmap = _interopRequireDefault(require("./tdtmap"));
var _tmap = _interopRequireDefault(require("./tmap"));
var _utils = require("./utils");
Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _utils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _utils[key];
    }
  });
});
