// @antv/l7-map/lib/handler/shim/dblclick_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
/**
 * The `DoubleClickZoomHandler` allows the user to zoom the map at a point by
 * double clicking or double tapping.
 */
class DoubleClickZoomHandler {
  /**
   * @private
   */
  constructor(clickZoom, TapZoom) {
    (0, _defineProperty2.default)(this, "clickZoom", void 0);
    (0, _defineProperty2.default)(this, "tapZoom", void 0);
    this.clickZoom = clickZoom;
    this.tapZoom = TapZoom;
  }

  /**
   * Enables the "double click to zoom" interaction.
   *
   * @example
   * map.doubleClickZoom.enable();
   */
  enable() {
    this.clickZoom.enable();
    this.tapZoom.enable();
  }

  /**
   * Disables the "double click to zoom" interaction.
   *
   * @example
   * map.doubleClickZoom.disable();
   */
  disable() {
    this.clickZoom.disable();
    this.tapZoom.disable();
  }

  /**
   * Returns a Boolean indicating whether the "double click to zoom" interaction is enabled.
   *
   * @returns {boolean} `true` if the "double click to zoom" interaction is enabled.
   */
  isEnabled() {
    return this.clickZoom.isEnabled() && this.tapZoom.isEnabled();
  }

  /**
   * Returns a Boolean indicating whether the "double click to zoom" interaction is active, i.e. currently being used.
   *
   * @returns {boolean} `true` if the "double click to zoom" interaction is active.
   */
  isActive() {
    return this.clickZoom.isActive() || this.tapZoom.isActive();
  }
}
exports.default = DoubleClickZoomHandler;
